### DISCUSSION TITLE
High CPU Utilization and Frequent Restarts after Upgrading from Keycloak 19 to Keycloak 22

### DISCUSSION THREAD

**Discussion by atulchauhan01** [keycloak/keycloak#30034] [Q&A]
We recently upgraded from Keycloak 19 to Keycloak 22 and have observed significant performance degradation in our system. Specifically, CPU utilization has spiked to 90% under normal load. Our setup includes a 2-node cluster running in ECS Fargate, and we’ve been experiencing frequent container restarts due to the error: **Terminating due to java.lang.OutOfMemoryError: Java heap space.**

Here are the steps we’ve taken to address the issue:

1. **Increased Heap Space**: We adjusted the initial and maximum memory settings using the following parameters:
`
-Xms1024m -Xmx3072m
`
However, this did not resolve the issue. The high CPU consumption and container restarts persisted.

2. **Memory Management Adjustments**: We noticed changes in memory management parameters in the Keycloak 24 kc.sh file and applied similar settings to our Keycloak 22 image:
`
-XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:MaxRAMPercentage=70 -XX:MinRAMPercentage=70 -XX:InitialRAMPercentage=50
`
Unfortunately, these adjustments also did not mitigate the problem. The CPU usage remained high, and the container continued to restart.
3. **Heap Dump Analysis**: We enabled heap dumps on OutOfMemoryError by setting -XX:+HeapDumpOnOutOfMemoryError and analyzed the heap dumps using Eclipse Memory Analyzer. Our analysis indicated that Freemarker is consuming most of the memory.
Attached are the logs and screenshots from the Eclipse Memory Analyzer for your reference.
![image](https://github.com/keycloak/keycloak/assets/6897456/cf831045-ba70-4cc2-bf63-9c750c1422cf)

![image](https://github.com/keycloak/keycloak/assets/6897456/79091b15-79ed-4c17-9d95-50e7308b7a03)

![image](https://github.com/keycloak/keycloak/assets/6897456/7f8b4a7a-0f5a-4ac5-886c-eff170201d22)

We would greatly appreciate any insights or suggestions on how to address these issues. Has anyone else experienced similar problems after upgrading to Keycloak 22? Are there specific configurations or optimizations that we might be missing?

**Comment by codespearhead** [comment]
Is this reproducible on the latest major release of Keycloak?

Possibly related: #27244 .

**Reply by atulchauhan01** [reply]
We are using Keycloak 22, not the latest version (Keycloak 24). We did not try with latest version so not sure if it is reproducible with latest version.

This issue (https://github.com/keycloak/keycloak/issues/27244) doesn't seem related. Although there was an issue with OpenShift support being discontinued in Keycloak, we are not running our cluster on Kubernetes.

We are not experiencing continuous **OutOfMemoryError**. The error message ("**Terminating due to java.lang.OutOfMemoryError: Java heap space.**") appears once before the container terminates. Additionally, CPU utilization is consistently high even with minimal usage. As you can see in the shared logs, this heap issue is caused by Freemarker.

**Reply by jonkoops** [reply]
We've had reports of this for Keycloak 22 before. We no longer actively support version 22, so the recommendation here is ti upgrade to the latest.
